pkg/prom/prom.go

```diff
@@ -198,6 +251,34 @@ func (rlpc *RateLimitedPrometheusClient) worker() {
 			roundTripStart := time.Now()
 			res, body, warnings, err := rlpc.client.Do(ctx, req)
 
+			// If retries on rate limited response is enabled:
+			// * Check for a 429 StatusCode OR 400 StatusCode and message containing "ThrottlingException"
+			// * Attempt to parse a Retry-After from response headers (common on 429)
+			// * If we couldn't determine how long to wait for a retry, use 1 second by default
+			if retryRateLimit {
+				var status []*RateLimitResponseStatus
+				var retries int = 5
+
+				for httputil.IsRateLimited(res, body) && retries > 0 {
+					retries--
+
+					// calculate amount of time to wait before retry, default to 1s
+					retryAfter := httputil.RateLimitedRetryFor(res, time.Second)
+					status = append(status, &RateLimitResponseStatus{RetriesRemaining: retries, WaitTime: retryAfter})
+					log.DedupedInfof(50, "Rate Limited Prometheus Request. Waiting for: %.2f seconds. Retries Remaining: %d", retryAfter.Seconds(), retries)
+
+					// execute wait and retry
+					time.Sleep(retryAfter)
+					res, body, warnings, err = rlpc.client.Do(ctx, req)
```

Comment:
To make this a bit more fault tolerant, there are other situations that would benefit from a retry. In our testing, we used https://github.com/hashicorp/go-retryablehttp defaults and added the weird ThrottleException check. See https://github.com/hashicorp/go-retryablehttp/blob/master/client.go#L435 as an example of what you might want to add to the retry.

Reply:
@dahamr Thanks for the added insight here. This HTTP client we're wrapping with the concurrent throttle is an  older version of the prometheus client (which, to be honest, I am not a fan of at all). I like the idea of a base retry policy - thanks for the reference. I can easily add the retry default timeout option, backoff, and max. However, I may hold off on the policy until we have some time to test further -- I'd really like to replace the prometheus client with a standard http client anyways - I've starred the retryable http client for later :)